content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+---
+title: "Towards Continuous Performance Regression Testing"
+date: 2020-12-16T17:45:00+01:00
+draft: false
+---
+:source-highlighter: rouge
+:rouge-style: base16.dark
+:icons: font
+:imagesdir: /images
+ifdef::env-github[]
+:imagesdir: ../../static/images
+endif::[]
+
+Functional unit and integration tests are a standard tool of any software development organization,
+helping not only to ensure correctness of newly implemented code,
+but also to identify regressions -- bugs to existing functionality introduced by a code change.
+The situation looks different though when it comes to regressions related to non-functional requirements, in particular performance-related ones:
+How to detect increased response times in a web application?
+How to identify decreased throughput?
+
+These aspects are typically hard to test in an automated and reliable way in the development workflow,
+as they are dependent on the underlying hardware and the workload of an application.
+For instance assertions on the duration of specific requests of a web application cannot be run in a meaningful way on a developer laptop which differs from the actual production hardware.
+When run in a virtualized or containerized CI environment, such tests are prone to failures due to concurrent load of other applications and jobs.
+
+This post introduces https://github.com/gunnarmorling/jfrunit[JfrUnit], which offers a fresh angle to this topic by supporting assertions not on metrics like latency/throughput themselves, but on _indirect metrics_ which may impact those.
```

Comment:
+100 awesome

But could use an example of "indirect metrics" and why they matter?

Reply:
It's clarified in the next sentence. Also reworded it a bit to make the point stronger.